sleap_io/io/nwb.py

```diff
@@ -0,0 +1,282 @@
+from pynwb import NWBFile, NWBHDF5IO, ProcessingModule
+from ndx_pose import PoseEstimationSeries, PoseEstimation
+
+
+from sleap_io import PredictedInstance, Labels, Video
+
+
+def _extract_predicted_instances_data(labels: Labels) -> pd.DataFrame:
+    """Auxiliar function to structure the predicted instances' data for nwb write.
+
+    Args:
+        labels (Labels): A general label object.
+
+    Raises:
+        ValueError: If no frames in the label objects contain predicted instances.
+
+    Returns:
+        pd.DataFrame: A pandas data frame with the structured data with
+        hierarchical columns. The column hierarchy is:
+                "video_path",
+                "skeleton_name",
+                "track_name",
+                "node_name",
+        And it is indexed by the frames.
+    """
+
+    # Form pairs of labeled_frames and predicted instances
+    labeled_frames = labels.labeled_frames
+    all_frame_instance_tuples = (
+        (label_frame, instance)
+        for label_frame in labeled_frames
+        for instance in label_frame.predicted_instances
+    )
+
+    # Extract the data
+    data_list = list()
+    for labeled_frame, instance in all_frame_instance_tuples:
+        # Traverse the nodes of the instances's skeleton
+        skeleton = instance.skeleton
+        for node in skeleton.nodes:
+            row_dict = dict(
+                frame_idx=labeled_frame.frame_idx,
+                x=instance.points[node].x,
+                y=instance.points[node].y,
+                score=instance.points[node].score,
+                node_name=node.name,
+                skeleton_name=skeleton.name,
+                track_name=instance.track.name if instance.track else "untracked",
+                video_path=labeled_frame.video.filename,
+            )
+            data_list.append(row_dict)
+
+    if not data_list:
+        raise ValueError("No predicted instances found in labels object")
+
+    labels_df = pd.DataFrame(data_list)
+
+    # Reformat the data with columns for dict-like hierarchical data access.
+    index = [
+        "skeleton_name",
+        "track_name",
+        "node_name",
+        "video_path",
+        "frame_idx",
+    ]
+
+    labels_tidy_df = (
+        labels_df.set_index(index)
+        .unstack(level=[0, 1, 2, 3])
+        .swaplevel(0, -1, axis=1)  # video_path on top while x, y score on bottom
+        .sort_index(axis=1)  # Better format for columns
+        .sort_index(axis=0)  # Sorts by frames
+    )
+
+    return labels_tidy_df
+
+
+def write_labels_to_nwb(labels: Labels, nwbfile_path: str):
+    """Write labels to an nwb file and save it to the nwbfile_path given
+
+    Args:
+        labels (Labels): A general label object
+        nwbfile_path (str): The path where the nwb file is to be written
+    """
+
+    session_description: str = "Processed SLEAP pose data"
+    session_start_time = datetime.datetime.now(datetime.timezone.utc)
+    identifier = "identifier"
```

Comment:
Maybe best to have these as keyword args with defaults for flexibility?

Reply:
Yep, good idea in case they need to change.